sound/soc/sof/ipc4.c

```diff
@@ -0,0 +1,218 @@
+	{115, "Reserved (ADSP_IPC_UNLOAD_INST_EXISTS removed)"},
+	{116, "Invalid target code ID provided"},
+	{117, "Injection DMA buffer is too small for probing the input pin"},
+	{118, "Extraction DMA buffer is too small for probing the output pin"},
+	{120, "Invalid ID of configuration item provided in TLV list"},
+	{121, "Invalid length of configuration item provided in TLV list"},
+	{122, "Invalid structure of configuration item provided"},
+	{140, "Initialization of DMA Gateway failed"},
+	{141, "Invalid ID of gateway provided"},
+	{142, "Setting state of DMA Gateway failed"},
+	{143, "DMA_CONTROL message targeting gateway not allocated yet"},
+	{150, "Attempt to configure SCLK while I2S port is running"},
+	{151, "Attempt to configure MCLK while I2S port is running"},
+	{152, "Attempt to stop SCLK that is not running"},
+	{153, "Attempt to stop MCLK that is not running"},
+	{160, "Reserved (ADSP_IPC_PIPELINE_NOT_INITIALIZED removed)"},
+	{161, "Reserved (ADSP_IPC_PIPELINE_NOT_EXIST removed)"},
+	{162, "Reserved (ADSP_IPC_PIPELINE_SAVE_FAILED removed)"},
+	{163, "Reserved (ADSP_IPC_PIPELINE_RESTORE_FAILED removed)"},
+	{165, "Reserved (ADSP_IPC_PIPELINE_ALREADY_EXISTS removed)"},
+};
+
+void sof_ipc4_check_reply_status(struct snd_sof_dev *sdev, u32 msg)
+{
+	u32 status = msg & SOF_IPC4_REPLY_STATUS_MASK;
+	int i;
+
+	sdev->ipc->msg.reply_error = status;
+	if (status) {
+		for (i = 0; i < ARRAY_SIZE(ipc4_status); i++) {
+			if (ipc4_status[i].status == status) {
+				dev_err(sdev->dev, "FW reported error: %s", ipc4_status[i].msg);
+				return;
+			}
+		}
+
+		dev_err(sdev->dev, "FW reported unknown error, status = %d", status);
+	}
+
+	snd_sof_ipc_reply(sdev, msg);
+}
+EXPORT_SYMBOL(sof_ipc4_check_reply_status);
+
+/* wait for IPC message reply */
+static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
+			     void *reply_data)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	int ret;
+
+	/* wait for DSP IPC completion */
+	ret = wait_event_timeout(msg->waitq, msg->ipc_complete,
+				 msecs_to_jiffies(sdev->ipc_timeout));
+	if (ret == 0) {
+		dev_err(sdev->dev, "error: ipc timed out for header:0x%x extension:0x%x",
+			msg->header, msg->extension);
+		return -ETIMEDOUT;
+	} else if (msg->reply_error) {
+		dev_err(sdev->dev, "error: ipc error for msg 0x%x : 0x%x\n",
+			msg->header, msg->extension);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sof_ipc4_tx_message_unlocked(struct snd_sof_ipc *ipc, u32 header,
+					u32 extension, void *msg_data, size_t msg_bytes,
+					void *reply_data, size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = ipc->sdev;
+	struct snd_sof_ipc_msg *msg;
+	int ret;
+
+	if (ipc->disable_ipc_tx)
+		return -ENODEV;
+
+	/*
+	 * The spin-lock is also still needed to protect message objects against
+	 * other atomic contexts.
+	 */
+	spin_lock_irq(&sdev->ipc_lock);
+
+	/* initialise the message */
+	msg = &ipc->msg;
+
+	msg->header = header;
+	msg->extension = extension;
```

Comment:
Let's clarify this again..
In IPC3 the `header` _is_ the `hdr->cmd` - the second 32 bit word from msg_data.
In IPC4 neither the `header`, nor the `extension` is stored within the msg_data, is this correct?

Reply:
sure, agree.             In my opinion, header was used in IPC3 but some patches removed the usage.

Reply:
> sure, agree. In my opinion, header was used in IPC3 but some patches removed the usage.

@RanderWang, as far as I can see, the `header` was always set to `hdr->cmd` without any exception. It is still used in one place instead directly using the `hdr->cmd`.

My question is still:
In IPC3 the `header` is the `hdr->cmd` - the second 32 bit word from msg_data.
In IPC4 neither the `header`, nor the `extension` is stored within the msg_data, is this correct?

Reply:
@ujfalusi yes, this is correct.